proposals/0000-pure-template-haskell.md

```diff
@@ -0,0 +1,294 @@
+The code in an impure splice must have type `Q a`, as with normal splices today. 
+
+The behaviour of normal splices is controlled with a new language extension, `EnforceTemplateHaskellPurity`. 
+The effect of `EnforceTemplateHaskellPurity` is that the code in a top-level normal splice must have type `PureQ a` instead of `Q a`.
+
+Additionally, there is a new language extension `PureTemplateHaskell`, which implies `NoImpureTemplateHaskell` and `EnforceTemplateHaskellPurity`.
+
+### 4. Changing the defaults and warnings
+
+At some point in the future we would like to turn on `EnforceTemplateHaskellPurity` by default. 
+To aid in this, we add a warning `-Wimpure-splices` (added to `-Wcompat`) which will warn if  `EnforceTemplateHaskellPurity` is not enabled, but a top-level splice would be rejected by it. 
+That is, if there is a normal splice which cannot be typed as `PureQ a`.
+
+TODO: it’s not entirely clear how we would implement this, we are complaining about a “hypothetical” type check. 
+
+The result of this is that to be `-Wcompat`-free a module with impure TH must a) enable `EnforceTemplateHaskellPurity` and `ImpureTemplateHaskell`, and b) use impure splice syntax. 
+In particular, this means that the module will continue to compile when we change the defaults (although `EnforceTemplateHaskellPurity` will become redundant).
+
+### 5. Proposed migration path
+
+The ideal migration path would go like this:
+
+- Implement the technical parts of this proposal
+- Users with `-Wcompat` will start to migrate to explicit impure splices.
+- Once much of the ecosystem is `Wimpure-splices`-free, switch `EnforceTemplateHaskellPurity` on by default.
+
+In parallel, as soon as there is significant explicit usage of `ImpureTemplateHaskell` in the wild, this will lead to pressure on library authors to purify the Template Haskell functions that they provide (since Haskellers will hate having something saying “impure” at the top of their files).
+
+This in turn will lead to pressure to ensure that `PureQ` has enough capabilities to cover major use cases (see “Supporting sufficiently many use cases” below).
+
+However, this process does not block changing the defaults, and can continue afterwards.
+
+## Examples
+
+Examples of `EnforceTemplateHaskellPurity`. 
+We're ignoring the stage restriction here for compactness, since it’s orthogonal to this proposal.
+
+```
+-- GOOD: Generating syntax definitions is entirely pure, but typically requires Reify
+$makeLenses ‘Foo
+
+-- BAD: reading a file requires IO, ‘embedFile’ has type ‘FilePath -> Q Exp’
+myfile1 = $(embedFile "dirName/fileName")
+-- GOOD: explicitly uses an impure splice
+myfile2 = $!(embedFile “dirName/fileName”)
+
+-- BAD: the inner splice is impure, leading to a MonadIO constraint on the overall splice, 
+-- which cannot be satisfied by a top-level pure splice
+myLazyfile3 = $(BSL.toLazy $(embedFile “dirName/fileName”))
```

Comment:
Perhaps 

```suggestion
myLazyfile3 = $(BSL.toLazy $!(embedFile “dirName/fileName”))
```

Reply:
Hmm, but now I'm not so sure if we should reject this! Impurity is isolated in `$!(...)`, so the overall splice should be OK. Shouldn't it? Can you clarify?

Reply:
As discussed above, only top-level splices use the new syntax, nested splices just propagate constraints out as they do today. c.f. we don't insist that a nested splice has type `Q a`, only top-level splices get that constraint.